include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,229 @@
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_DECODED_VIEW_HPP
+#define BOOST_URL_IMPL_DECODED_VIEW_HPP
+
+namespace boost {
+namespace urls {
+
+class decoded_view::iterator
+{
+    char const* begin_{nullptr};
+    char const* pos_{nullptr};
+    char const* end_{nullptr};
+    bool plus_to_space_{true};
+
+    friend decoded_view;
+
+    iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space) noexcept
+        : begin_(str)
+        , pos_(str)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    // end ctor
+    iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    iterator() = default;
+
+    iterator(iterator const&) = default;
+
+    iterator&
+    operator=(iterator const&) = default;
+
+    reference
+    operator*() const noexcept
+    {
+        char c{'\0'};
+        pct_decode_opts opt;
+        opt.plus_to_space = plus_to_space_;
+        pct_decode_unchecked(
+            &c, &c+1, string_view(pos_, end_), opt);
+        return c;
+    }
+
+    iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        if (*pos_ != '%')
+            ++pos_;
+        else
+            pos_ += 3;
+        return *this;
+    }
+
+    iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ < 3 ||
+            pos_[-3] != '%')
+            --pos_;
+        else
+            pos_ -= 3;
+        return *this;
+    }
+
+    iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    const char*
+    base()
+    {
+        return pos_;
+    }
+
+    bool
+    operator==(
+        iterator const& other) const noexcept
+    {
+        return pos_ == other.pos_;
+    }
+
+    bool
+    operator!=(
+        iterator const& other) const noexcept
+    {
+        return !(*this == other);
+    }
+};
+
+inline
+auto
+decoded_view::
+begin() const noexcept ->
+    decoded_view::const_iterator
+{
+    return {p_, n_, plus_to_space_};
+}
+
+inline
+auto
+decoded_view::
+end() const noexcept ->
+    decoded_view::const_iterator
+{
+    return {p_, n_, plus_to_space_, 0};
+}
+
+inline
+decoded_view::const_reference
```

Comment:
could use `auto` here to skip the namespace qualifier

Reply:
or use a trailing return type.